pytorch_pfn_extras/training/extensions/slack.py

```diff
@@ -0,0 +1,205 @@
+        end_text_template (str or callable): Template for sending a message
+            at the completion of the experiment.
+            See ``text_template`` for format.
+        filenames_template (list of str or callable): list of files that will
+            be uploaded to slack, these are string templates that can take
+            values in the same way as ``text_template``. Optional.
+        use_threads (bool): If subsequent calls of this extensions should be
+            posted as a thread of the original message or not.
+            Default is ``False``.
+        context_object (object): Custom object that contains data used to
+            format the text or the filenames. Optional, default is ``None``.
+        webhook_url (str): Webhook to send messages, it is mutually exclusive
+            with ``token`` and ``client``. Used when slack_sdk is not available.
+            Optional, default is ``None``.
+        token (str): Token for the slack api, if ``None`` the environment
+            variable ``SLACK_TOKEN`` will be used. Ignored if ``client`` is
+            supplied. Optional, default is ``None``.
+        client (slack_sdk.WebClient): In case that there is an already created
+            slack client in the application, allows to directly use it.
+            Optional, default is ``None``
+        trigger: Trigger that decides when to send the messages.
+            This is distinct from the trigger of this extension
+            itself. If it is a tuple in the form ``<int>, 'epoch'`` or
+            ``<int>, 'iteration'``, it is passed to :class:`IntervalTrigger`.
+    """
+    def __init__(
+        self,
+        channel_id: str,
+        text_template: Union[
+            str, Callable[[ExtensionsManagerProtocol, Any, dict], str]
+        ],
+        start_text_template: Optional[Union[
+            str, Callable[[ExtensionsManagerProtocol, Any, dict], str]
+        ]] = None,
+        end_text_template: Optional[Union[
+            str, Callable[[ExtensionsManagerProtocol, Any, dict], str]
+        ]] = None,
+        filenames_template: Optional[
+            List[
+                Union[
+                    str, Callable[[ExtensionsManagerProtocol, Any, dict], str]
+                ]
+            ]
+        ] = None,
+        use_threads: bool = False,
+        context_object: Optional[object] = None,
+        webhook_url: Optional[str] = None,
+        token: Optional[str] = None,
+        client: Optional[Any] = None,  # slack_sdk.WebClient, Any to avoid mypy errors
+        trigger: trigger_module.TriggerLike = (1, 'epoch'),
+    ) -> None:
+        if not _slack_sdk_available and webhook_url is None:
+            warnings.warn(
+                '`slack_api` package is unavailable. '
+                'Slack will do nothing unless a webhook url is provided.')
+            return
+
+        if not _requests_available and webhook_url is not None:
+            warnings.warn(
```

Comment:
Can we use urllib instead of requests to avoid introducing additional dependency?